abstract/encoding.go

```diff
@@ -170,13 +193,31 @@ func (de *decoder) value(v reflect.Value, depth int) error {
 			}
 		}
 
-	case reflect.Slice, reflect.Array:
+	case reflect.Slice:
+		if v.IsNil() {
+			panic("slices must be initialized to correct length before decoding")
+		}
+		fallthrough
+	case reflect.Array:
 		l := v.Len()
 		for i := 0; i < l; i++ {
 			if err = de.value(v.Index(i), depth+1); err != nil {
 				return err
 			}
 		}
+
+	case reflect.Int:
```

Comment:
Since 64-bit architectures are increasingly the norm, I think we should define 64-bit as the standard "wire-format size" of an int, and just produce a decoding error on 32-bit platforms if the decoded int is too big.  I'm committing a change to do this.